DOCS-2142-4.x-dataweave-memory-management-kt (#949)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-memory-management.adoc

> endif::[]

Reviewer: ORG
= DataWeave Memory Management
== RAM vs Disk Usage

Because there is only one subtopic, do you suppose we should reword the container title so that it encompasses that solitary subt, and omit the division? "Fine-Tune Large File Memory Management"? "Manage DataWeave RAM and Disk Storage"?

Or, we could simply omit the subtitle and include RAM vs. disk use in the topic intro text:

When processing large files through DataWeave in Mule runtime engine, consider the availability of RAM versus disk storage so that you can fine-tune how much memory is used, and when.

:keywords: studio, anypoint, esb, transform, transformer, format, aggregate, rename, split, filter convert, xml, json, csv, pojo, java object, metadata, dataweave, data weave, datamapper, dwl, dfl, dw, output structure, input structure, map, mapping, streaming, weaveoutputhandler

> When processing large files through DataWeave in Mule runtime engine, there are a few things you can set up to fine-tune how much memory will be used and when.

Reviewer: See suggestion above. Otherwise:
```suggestion
To fine-tune processing large files through DataWeave in Mule runtime engine, you can set controls for how much memory is used, and when.
```

[[buffersize]]
== RAM vs Disk Usage

> DataWeave uses disk for storing information and not running out of memory. The files that are going to be created are going to be placed on the temporary directory. If you desire a custom directory instead of the system default, you can specify the directory using the `java.io.tmpdir` property.

Reviewer: ```suggestion
DataWeave uses disk data storage to avoid running out of memory. The data files created are placed in a default temporary directory. If you want to store those files in a custom directory instead, you can specify the directory by using the `java.io.tmpdir` property.
```
BUT...can we say something about WHY the audience might want or need to create a custom directory?

> All DataWeave files are going to have the prefix `dw-buffer-` and there are two kind of files that are being generated. +

Reviewer: ```suggestion
Two types of DataWeave data files are generated, both with names beginning with `dw-buffer-`:+
```

Reviewer: BUT...how does this text relate to the paragraph above it? What do the DataWeave buffer file names have to do with running out of memory or file placement in a default or custom directory?
You might need a peer review to help with the wording and flow of this topic.

> - `dw-buffer-output-${count}.tmp` is used to store the output of a transformation when the result is bigger than the threadshold *1572864 bytes*.

Reviewer: ```suggestion
- `dw-buffer-output-${count}.tmp`
Used to store the output of a transformation when the result exceeds the threshold of 1572864 bytes.
```

Reviewer: How does listing the two file names relate to adding system properties to change the threshold?

> To change the threshold value at which memory is no longer used as a buffer, you must add the system property `com.mulesoft.dw.buffersize` and assign it the number (in bytes) of your new threshold.  System properties may be defined in several ways. See xref:configuring-properties.adoc#system-properties[system properties] for more details on how to set them.

Reviewer: ```suggestion
To change this threshold value, add the system property `com.mulesoft.dw.buffersize` and assign it the number of bytes you want as your new threshold. Because you can define system properties in several ways, see xref:configuring-properties.adoc#system-properties[system properties] for further details.
```

> Mule runtime engine deletes the file when the value is no longer referenced, JVM GC collects it or when the Mule Event finishes executing.

Reviewer: Review this sentence, because this part is not clear: "...referenced, JVM GC collects it or when the Mule Event...."

Also, what causes the file value to be no longer referenced? There seem to be some conceptual gaps in the narrative flow.

> - `dw-buffer-index-${count}.tmp` is used to store index information of a value being read. This file helps DataWeave access data quickly. Mule runtime engine deletes the file when the execution of the transformation ends or in the case of a streaming use case like the foreach, when the stream ends, meaning when the foreach finishes its execution.

Reviewer: ```suggestion
* `dw-buffer-index-${count}.tmp`
Used to store index information of a value being read. This file helps DataWeave access data quickly. Mule runtime engine deletes the file when the execution of the transformation ends or, in a streaming use case like the foreach loop, when the stream ends (when foreach finishes its execution).
```
